src/connection/connection.rs

```diff
@@ -72,25 +72,18 @@ impl Connection {
         let init_addresses = init_addresses.iter().map(|addr| addr.as_ref().parse()).try_collect()?;
         let addresses = Self::fetch_current_addresses(background_runtime.clone(), init_addresses, credential.clone())?;
 
-        // FIXME: switch to connecting to all servers (not at least one) when the issues with auth in Cluster will be fixed (PR#482)
-        let mut server_connections = HashMap::new();
-        let mut error_buffer = Vec::new();
+        let mut server_connections = HashMap::with_capacity(addresses.len());
         for address in addresses {
             let server_connection =
-                ServerConnection::new_encrypted(background_runtime.clone(), address.clone(), credential.clone());
-            match server_connection {
-                Ok(server_connection) => {
-                    server_connections.insert(address, server_connection);
-                }
-                Err(err) => {
-                    error_buffer.push(format!("- {}: {}", address, err));
-                }
-            }
+                ServerConnection::new_encrypted(background_runtime.clone(), address.clone(), credential.clone())?;
+            server_connections.insert(address, server_connection);
         }
-        if server_connections.is_empty() {
-            Err(ConnectionError::ClusterAllNodesFailed(error_buffer.join("\n")))?
+        let (ok, errors): (Vec<_>, Vec<_>) = server_connections.clone().into_iter().map(|(_addr, conn)| conn.validate()).partition(Result::is_ok);
+        if ok.is_empty() {
+            Err(ConnectionError::ClusterAllNodesFailed(errors.into_iter().map(|err| err.unwrap_err().to_string()).collect::<Vec<_>>().join("\n")))?
+        } else {
+            Ok(Self { server_connections, background_runtime, username: Some(credential.username().to_string()) })
         }
```

Comment:
That's a cool construct, but it can be cooler.

1) `clone` just to validate? We should be able to just use a reference to the server connection here.
2) We don't actually use the `ok` partition, so it can just be `filter_map(Result::err).collect()`.
3) `_` shouldn't really be used in the type hints, write the error type out.